[4.0] Subform UX

joomdonation opened the pull request:
Pull Request for Issue #28389.

### Summary of Changes
Joomla! has a custom field type called **Subfields** and recently, renamed to **Subform** in a different PR. From it's documentation [https://docs.joomla.org/Custom_fields_type:_Subfields](https://docs.joomla.org/Custom_fields_type:_Subfields)

> the aim of it is to provide the user the possibility to create a repeatable custom field for an item with multiple related fields

There are some UX issues with that kind of custom field as described below (see https://github.com/joomla/joomla-cms/issues/28389 for detailed discussion):

- On Fields Management screen, there is no way to know if a custom field is a standard field or it is a subfield (part of a **Subform** custom field type).
- To prevent a custom field from being used both as standard field and as part of a **Subform**, we need to set **Category** of that custom field to **None**. It's hard to know about that.
- There is no filter option to allow seeing custom fields which are being used as part of **Subform**.
- There is no option to limit maximum number of rows which can be added when adding data for a **Subform**. 
- Users won't know that they must add some standard fields before they can add a **Subform**.

This PR tries to solve these issues base on suggestions from @AndySDH. See https://github.com/joomla/joomla-cms/issues/28389#issuecomment-816878826 for details (one of the suggestion is renaming the custom field type from Subfields to Subform but it is handled by different PR) 

1. Add a new setting **Use In Subform** on add/edit custom field screen to indicate that the field will be used in **Subform**.

- When it set to Yes, categories selection will become hidden.
- Category of the field will become **None**.

2. Add a filter to search tool to allow filtering fields belong to Subform (Yes/No). By default, all fields will be show.

<img width="893" alt="filter" src="https://user-images.githubusercontent.com/977664/114297720-7f0d6900-9adc-11eb-98e1-dbed55df5703.png">

3. If you add a **Subform** custom field but forgot to create at least one standard field before , there will be a warning displayed at the top to tell you about that.
![warning](https://user-images.githubusercontent.com/977664/114297790-efb48580-9adc-11eb-82ca-5dc5d85b1165.png)

4. On **Custom Fields Management** screen, add a small badge next to the field so that we know if it is a field which is used for **SubForm Only**.

![subform_only](https://user-images.githubusercontent.com/977664/114305273-a296da80-9b01-11eb-9331-b269f447a2b1.png)

5. Add a new setting called **Maximum Rows** to control maximum number of rows which you can add while adding data for that **Subform** custom field type.

### Testing Instructions
1. Install Joomla 4 nightly build. Or if you are having an existing Joomla 4 installation, please update that installation to latest Joomla 4 nightly build.
2. Download update package for this PR at https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/33096/downloads/42296/Joomla_4.0.0-beta8-dev+pr.33096-Development-Update_Package.zip , go to **System -> Update -> Joomla**, upload and install that update package
3. Go to Content -> Custom Fields, create one or two **Text** custom fields. On adding custom field screen, confirm that you see a new setting called **Only Use In Subform**. Set it to **Yes** and confirm that the **Category** selection at the right is hidden. Save the field, you will be redirected to Custom Fields list screen. Confirm that you see a badge **Subform Only** displays next to that custom field (to indicate that the custom field is only used in Subform)
4. Create another custom field type **Subform** . Confirm that:
- You do not see **Only Use In Subform** setting.
- You see a new setting called **Maximum Rows**. Set it to the number you want (for example 2). In the **Fields** setting of that custom field, select some fields for it.
5. Go back to custom fields list screen, click on **Filter Options** to show search tools. You will see a new filter option **Only Use In Subform**, use it and make sure it is working as expected
6. Now, create a new article. Navigate to Fields tab, enter data for that Subform custom field which you created. Please make sure that you cannot add more rows for that custom field than the number you entered into **Maximum Number Rows** setting of that field (2 in the case)

joomdonation committed:
Draft SubFields UX

AndySDH commented:
This looks nice @joomdonation, thanks for working on this!

1. That setting should be called "**Only** Use in Subform".
As you are normally already free to use a field both as a normal field (so with category assignments) AND inside a subform.
So since that toggle will make the Category Input hide, it should be used for fields you want to be used ONLY in subforms and nowhere else.

2. Great!

3. Nice! What happens at the bottom part of the screen, where the Field Selection Dropdown would be? Does it still show an empty dropdown?

joomdonation commented:
1. Easy. It is just a language item which we need to change. I haven't added it to language file yet :D. As I said, quick prototype PR only

3. It is still as before, empty dropdown. I want to leave it here so that the system will prevent user from saving the field without having to enter any any other coding works, check.....

Do you have any idea to indicate a field is Use In SubForm only on custom fields list screen?

AndySDH commented:
> It is still as before, empty dropdown. I want to leave it here so that the system will prevent user from saving the field without having to enter any any other coding works, check.....

Makes sense. The warning at the top should suffice.

> Do you have any idea to indicate a field is Use In SubForm only on custom fields list screen?

Sure, nice idea!

We could hide a label like this (pardon the J3 interface, just a quick draft to convey the idea):

![image](https://user-images.githubusercontent.com/25923699/114298394-7bff8200-9ab6-11eb-90a7-fad4ca662681.png)

or alternatively, since Category is set to None anyway, even something like this:

![image](https://user-images.githubusercontent.com/25923699/114298412-99cce700-9ab6-11eb-97d6-deaa471ce2c4.png)

joomdonation committed:
SubForm Only Indicator

joomdonation commented:
Don't ask me for layout :D. I made a quick modification and now it looks like below 

![subform_only](https://user-images.githubusercontent.com/977664/114298623-8c792200-9ae1-11eb-8ba0-8dcb0c0d12c1.png)

AndySDH commented:
Looks good! I would probably put it on the side, rather than below. Similar to how "Hidden" is on the side for Hidden Menu Items. To optimize the available space.
Nice job :)

joomdonation commented:
I will ask someone to help with layout things :). Now the next thing we should discuss about migrating data. Somehow, we will have to mark existing fields which is used in subform as **Use In SubForm Only**. For existing fields, what fields we should do that change?

- Fields which are part of subform
- And assigned to Category None ?

Is that enough?

And do you know when subform was added to Joomla? I want to ask this because it is just added to Joomla 4, maybe we will have to update the script at https://github.com/joomla/joomla-cms/pull/32611 to do this conversion when migrating repeatable fields to subform fields from Joomla 3 to Joomla 4 only

AndySDH commented:
Yes I think that's more than enough. It's probably not even that necessary to do that as part of the migration, as they will work just fine even if not marked as "Subform Only". 

But if you wanna go the extra mile and do that as well also in the migrationion, the two checks you mentioned (Used in subform & Category none) are more than enough indeed :)

AndySDH commented:
Wait, let me take that back. Category "None" is not even part of J3. It's a J4 feature.

So I would say we shouldn't care about this in the migration.

joomdonation commented:
> Wait, let me take that back. Category "None" is not even part of J3. It's a J4 feature.
> 
> So I would say we shouldn't care about this in the migration.

Nice, less work for us :)

AndySDH commented:
Actually, let me take that back again (sorry, ideas keep coming to mind lol).

In Joomla 3, we only had "Repeatable" field. Joomla 4 introduces the new Subform/Subfields type.

So this means, that ALL Repeatable fields that might have been created in Joomla 3, are ALL actually intended to be used only in Subform (there was no way to use it as normal field as well).

So, all fields part of the J3 repeatable field make sense to be migrated to "Subform Only", if you want to do that, as part of the migration process.

joomdonation commented:
@brianteeman Could I ask you for suggestions about two terms:

- The setting which is used to determine if a field is used on SubForm only. Can we use **Only use in Subform** ? I need to know now because the name which use to store data in database table will depend on the terrm.
- The badge which is used to show on Custom Fields management screen to show that a field is used for SubForm only. Can we use the term **SubForm Only**?

joomdonation commented:
> Actually, let me take that back again (sorry, ideas keep coming to mind lol).
> 
> In Joomla 3, we only had "Repeatable" field. Joomla 4 introduces the new Subform/Subfields type.
> 
> So this means, that ALL Repeatable fields that might have been created in Joomla 3, are ALL actually intended to be used only in Subform (there was no way to use it as normal field as well).
> 
> So, all fields part of the J3 repeatable field make sense to be migrated to "Subform Only", if you want to do that, as part of the migration process.

I can update the migration script to have it works like that later after we finish this one. And just final question: With the change in this PR when it is ready, do you think we can close the release blocker https://github.com/joomla/joomla-cms/issues/28389 ?

AndySDH commented:
Cool!

#28389 still has other issues to be addressed, see my last comment, point 4 and 5. Also point 2 is important, we have to decide a final naming in #32864.

brianteeman commented:
### Some questions.
Why subform only and why no category?

### Scenario
I have the following fields
- name
- date
- age
- school
- class
- city

I have a subform called english with the following fields
- name
- age

I have another subform that is only displayed in the category exam with the following fields
- name
- school
- class

For me all I would be changing on the list view of all fields would be to indicate if a field is used in a subform and the name of the subform. ie exactly like category but in addition to and not instead of

Am I missing something that prevents that?

brianteeman commented:
oh and its not really a subform but a group/collection of fields

AndySDH commented:
@brianteeman As far as "Why Category None", it's because a field that is set to be "only used in a Subform" is not (cannot) be assigned to a category. Its category assignment will be solely related to its "Subform / Parent Field", so hence why its own category becomes "None".

AndySDH commented:
> oh and its not really a subform but a group/collection of fields

Yeah, hence why not sure if we should call it Subform in your PR? https://github.com/joomla/joomla-cms/pull/32864

joomdonation commented:
Actually, if you assign that field to a a category (not None), when you add an item (for example, an article), the field will be displayed both as standard field and as part of SubForm, so it is confusing (not sure if I understand the question correctly)

joomdonation commented:
But if we introduce a new setting **Use In SubForm Only** (call it anything we like), the field can still be assigned to categories if we want. Then on dropdown when we add new SubForm field, we will only show field which has **Use In SubForm Only** set to **Yes**. When display form to allow add/edit item, we can hide these fields, I think.

brianteeman commented:
I dont understand why "use in subform only" do you actually mean "this field is used in a subform called x"

AndySDH commented:
> I dont understand why "use in subform only" do you actually mean "this field is used in a subform called x"

It's just so to mark a field as usable ONLY in a subform, and not as normal field.

By default, you can use a field BOTH as part of a subform and a normal field.

So for example, you can have a field "Age" assigned to the "People" category as a normal field. And have it part of the "English" subform somewhere else.

By setting it as a "Subform only", you ensure the field is not normally assigned to any category, and only available for selection in Subforms.

joomdonation commented:
And if it is needed, when we add a new SubForm field type (called Subfields at the moment), we can only display fields marked as **use in subform only**

That would help choosing fields easier. Right now, it displays all fields in dropdown and in case you have many fields, It will be hard to choose the field

![limit_fields](https://user-images.githubusercontent.com/977664/114300040-7ae74880-9ae8-11eb-9a0c-0e90eac01733.png)

AndySDH commented:
> And if it is needed, when we add a new SubForm field type (called Subfields at the moment), we can only display fields marked as **use in subform only**
> 
> That would help choosing fields easier. Right now, it displays all fields in dropdown and in case you have many fields, It will be hard to choose the field

That depends. That would be a loss of functionallity. You would have to create SPECIFIC subform fields, instead of being able to choose from ANY field.

If we were do to that, then we would lose the ability to do this:

> you can use a field BOTH as part of a subform and a normal field.
> So for example, you can have a field "Age" assigned to the "People" category as a normal field. And have it part of the "English" subform somewhere else.

brianteeman commented:
I think you are overthinking this. Just let me see if a field is being used in a subform and the name of the form. Doesnt need anything more than that My 2c.

AndySDH commented:
> I think you are overthinking this. Just let me see if a field is being used in a subform and the name of the form. Doesnt need anything more than that My 2c.

The thing is, without a setting "Only use in Subform", you would need to remember to manually set a field category to "None" if you don't want to also appear as a normal field. Could be (is) confusing.

joomdonation commented:
@brianteeman In theory, a field could be used on multiple SubForm fields, and we do not have an easy way to know that it is used as part of subform at the moment, that's why we introduce a new setting

The ability to filter fields to see list of fields which are using is useful, too, I think

Finally, set Category to None automatically is good, too. See the reason in my earlier comment https://github.com/joomla/joomla-cms/pull/33096#issuecomment-817280985

So I would leave it works as how it is in this PR (beside change the terms if needed)

AndySDH commented:
By the way, other note, if we go this way, we can actually remove the ability to set a category to "None" in the dropdown itself. The only reason why it was introduced it was for this purpose.

If we have the "Only use in Subform?" setting, which hides the Category selection UX completely then the "None" selection in the category field is useless and makes no sense anymore to have :)

chmst commented:
I rarely use field, so I am kind of novice user here.

First of all I have problems with the terminology - subfields - sub-field - subform - as a user I don't understand the whole fields terminology. (Not in scope of this PR - Why do we have Field groupp, which is a field catgegory? I know it from develoeprs point of view, but not as a user). 

So my2cent: 

If I understand this PR  right, we have three groups of fields:
- not in a  subform
- used  in a subform or stand alone
- only used in a subform
It is a good idea to have a filter for that.

New  or edit:
Categories preset with "All" as it is now should be sufficient
An indicator for "used  in subforms" indicates that the field is  used in subforms. 

Overview:
- not in a  subform
- used  in a subform or stand alone - "used in subforms - button" as suggested

Categories:
- not in a  subform - show the categories
- used  in a subform or stand alone - "used in subforms aa, bb, .." , show the categories
- only used in a subform  -  "used in subforms: xx, aa, ..." , hide the category field

joomdonation committed:
Change field name, add language items...

joomdonation committed:
CS

joomdonation committed:
Add missing language item

joomdonation committed:
Use Subform instead of SubForm

joomdonation commented:
@AndySDH When you have time, could you please give this PR a quick test to see if I am missing anything else (base on your suggestions) before I mark this PR as ready for review? As it contains change to database, please download the update package for this PR from https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/33096/downloads/41974/ , then go to System -> Joomla -> Update to update the package to test.

Quy commented on administrator/language/en-GB/com_fields.ini:
```diff
@@ -46,6 +46,8 @@ COM_FIELDS_FIELD_N_ITEMS_TRASHED_1="Field trashed"
 COM_FIELDS_FIELD_N_ITEMS_UNPUBLISHED="%d fields unpublished"
 COM_FIELDS_FIELD_N_ITEMS_UNPUBLISHED_1="Field unpublished"
 COM_FIELDS_FIELD_NOTE_LABEL="Note"
+COM_FIELDS_FIELD_ONLY_USE_ON_SUBFORM_BADGE="Subform Only"
```
```suggestion
COM_FIELDS_FIELD_ONLY_USE_IN_SUBFORM_BADGE="Subform Only"
```

Quy commented on administrator/components/com_fields/tmpl/fields/default.php:
```diff
@@ -158,6 +158,11 @@
 											<?php endif; ?>
+											<?php if ($item->only_use_in_subform) : ?>
+												<div class="small badge bg-secondary">
+													<?php echo Text::_('COM_FIELDS_FIELD_ONLY_USE_ON_SUBFORM_BADGE'); ?>
```
```suggestion
													<?php echo Text::_('COM_FIELDS_FIELD_ONLY_USE_IN_SUBFORM_BADGE'); ?>
```

Quy commented on administrator/components/com_fields/forms/filter_fields.xml:
```diff
@@ -64,6 +64,18 @@
 		</field>
+
+		<field
+				name="only_use_in_subform"
+				type="list"
+				label="COM_FIELDS_FIELDS_FILTER_ONLY_USE_IN_SUBFORM_LABEL"
+				onchange="this.form.submit();"
+				default=""
+		>
```
```suggestion
			name="only_use_in_subform"
			type="list"
			label="COM_FIELDS_FIELDS_FILTER_ONLY_USE_IN_SUBFORM_LABEL"
			onchange="this.form.submit();"
			default=""
			>
```

Quy commented on administrator/components/com_fields/forms/field.xml:
```diff
@@ -76,6 +77,16 @@
 		</field>
 
+		<field
+				name="only_use_in_subform"
+				type="radio"
+				label="COM_FIELDS_FIELD_ONLY_USE_IN_SUBFORM_LABEL"
+				layout="joomla.form.field.radio.switcher"
+				default="0"
+		>
```
```suggestion
			name="only_use_in_subform"
			type="radio"
			label="COM_FIELDS_FIELD_ONLY_USE_IN_SUBFORM_LABEL"
			layout="joomla.form.field.radio.switcher"
			default="0"
			>
```

joomdonation commented:
@Quy Could you please commit your suggested changes to language/codestyle/layout please?

Commit:
Update administrator/language/en-GB/com_fields.ini

Commit:
Update administrator/components/com_fields/tmpl/fields/default.php

Commit:
Update administrator/components/com_fields/forms/filter_fields.xml

Commit:
Update administrator/components/com_fields/forms/field.xml

AndySDH commented:
@joomdonation 

> @AndySDH When you have time, could you please give this PR a quick test to see if I am missing anything else (base on your suggestions) before I mark this PR as ready for review? As it contains change to database, please download the update package for this PR from https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/33096/downloads/41974/ , then go to System -> Joomla -> Update to update the package to test.

Sure, I will take a look at this soon :)

joomdonation committed:
Add max rows parameter

joomdonation committed:
Fix typo

AndySDH commented:
Small thing that I can see from the code: the new max_rows parameters should only show if is set to 1

joomdonation committed:
Showon

joomdonation commented:
> Small thing that I can see from the code: the new 'max_rows' parameter should only show if 'repeat' is set to 1
> (it only applies if the field is repeatable)

Done !

AndySDH commented:
@joomdonation the download package doesn't work :(

joomdonation commented:
@AndySDH  I don't know if the SQL command is executed when we install the update package or not. Maybe you can run this SQL command manually to your database via phpmyadmin

ALTER TABLE `jos_fields` ADD `only_use_in_subform` tinyint(1) NOT NULL DEFAULT 0; 

(replace jos_ with the prefix of your site database)

joomdonation commented:
@richard67 Do you know how to to execute the SQL changes in a PR like this one?

richard67 commented:
@joomdonation They will be executed when someone updates to the update package built for this PR by drone, and they can be run "manually" in an SQL client like e.g. phpMyAdmin, and they can be run by applying the PR and then going to "System - Manage - Database" and using the fix button.

joomdonation commented:
@richard67 Thanks. This afternoon, I tried to download the update package for this PR from https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/33096/downloads/42053/ , and go to System -> Update -> Joomla, upload that package to update but it seems the SQL is not executed. I had to run the SQL command manually via phpmyadmin

richard67 commented on administrator/components/com_admin/sql/updates/mysql/4.0.0-2021-04-11.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE `#__fields` ADD `only_use_in_subform` tinyint(1) NOT NULL DEFAULT 0;
```
```suggestion
ALTER TABLE `#__fields` ADD COLUMN `only_use_in_subform` tinyint(1) NOT NULL DEFAULT 0;

```

richard67 replied:
Newline at the end of file missing. Key word "COLUMN" missing in statement. It's needed for the database checker/fixer.

richard67 commented on administrator/components/com_admin/sql/updates/postgresql/4.0.0-2021-04-11.sql:
```diff
@@ -0,0 +1,2 @@
+-- Add locked field to extensions table.
+ALTER TABLE "#__fields" ADD COLUMN "only_use_in_subform" smallint DEFAULT 0 NOT NULL;
```
```suggestion
ALTER TABLE "#__fields" ADD COLUMN "only_use_in_subform" smallint DEFAULT 0 NOT NULL;

```

richard67 replied:
Newline at the end of file missing.

richard67 commented:
@joomdonation Wait with committing my suggestions above ... I might have to correct them by another mistake.

AndySDH commented:
I tested all of the applied changes succesfully, great job @joomdonation !

However, I have a few notes for issues I encountered, or further improvements that should be made:

**FIELD CREATION FORM**

- Should we add a default to maximum number rows option? Now it's '0', should we maybe leave it empty? Not sure 0 looks nice.

**FIELDS LISTING**

- As mentioned before, I'd put the 'Subform Only' badge on the side of the field name, instead of under it (similar to how it's done for hidden Menu Items)

**ARTICLE EDIT FORM**

- I think we should surround the whole Subform with a fieldset, for clarity. When a field is not repeatable, it's not easy to tell the child fields apart from normal fields.

- A media file doesn't work inside a Subform. When you click "Select" to select a picture, nothing happens.

- The subform table doesn't occupy the 100% width of the space available in the backend edit form:
![Cattura](https://user-images.githubusercontent.com/25923699/115116681-25ed7b80-9f9b-11eb-8443-be7088ff0b63.PNG)

- The subform table exceeds the maximum width available in the frontend edit form, making it impossible to edit or add new rows.
![2Cattura](https://user-images.githubusercontent.com/25923699/115116682-28e86c00-9f9b-11eb-847c-67b498598b33.PNG)

**REPEATABLE MIGRATION FROM JOOMLA 3**

- All migrated child field from Joomla 3 should be marked as "only_use_in_subform" = 1

--------

Apart from these fixes, everything else works as explained!

joomdonation commented:
@richard67 I'm unsure what is the change? Feel free to commit it yourself

Quy commented on plugins/fields/subfields/params/subfields.xml:
```diff
@@ -18,7 +18,14 @@
 			</field>
-
+			<field
+					name="max_rows"
+					type="number"
+					label="PLG_FIELDS_SUBFIELDS_PARAMS_MAX_ROWS_LABEL"
+					default="0"
+					filter="integer"
+					showon="repeat:1"
```
```suggestion
				name="max_rows"
				type="number"
				label="PLG_FIELDS_SUBFIELDS_PARAMS_MAX_ROWS_LABEL"
				default="0"
				filter="integer"
				showon="repeat:1"
```

Commit:
Update plugins/fields/subfields/params/subfields.xml

joomdonation commented:
@AndySDH 

- For the change in article edit form, it should be handled in a separate PR, I think.  Please don't ask me to do the layout stuff, it would take me years for doing that :D. Someone with better frontend skill could help with that task

richard67 commented:
@joomdonation In SQL files we also have a newline at the end of the file. And in the MySQL file, the "COLUMN" was missing. It is also valid SQL syntax without it, but the database fixer will not understand it. Therefore the golden rule when creating update SQL scripts for Joomla is to look for an already existing script which does the same (adding a column) and stay with the syntax of that.

AndySDH commented:
> @AndySDH
> 
> * For the change in article edit form, it should be handled in a separate PR, I think.  Please don't ask me to do the layout stuff, it would take me years for doing that :D. Someone with better frontend skill could help with that task

Sure, just raising the issues :)

joomdonation commented:
@Quy Could you help me with this:

> As mentioned before, I'd put the 'Subform Only' badge on the side of the field name, instead of under it (similar to how it's done for hidden Menu Items)

joomdonation committed:
Update administrator/components/com_admin/sql/updates/postgresql/4.0.0-2021-04-11.sql

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

AndySDH commented on administrator/language/en-GB/com_fields.ini:
```diff
@@ -82,6 +85,8 @@ COM_FIELDS_GROUPS_FILTER_SEARCH_LABEL="Search Field Groups"
 COM_FIELDS_GROUPS_TABLE_CAPTION="Table of Field Groups"
 COM_FIELDS_MUSTCONTAIN_A_TITLE_FIELD="Field must have a title."
 COM_FIELDS_MUSTCONTAIN_A_TITLE_GROUP="Field Group must have a title."
+COM_FIELDS_NO_FIELDS_TO_CREATE_SUBFORM_FIELD_WARNING="You need to create some standard fields before creating Subform field"
```
```suggestion
COM_FIELDS_NO_FIELDS_TO_CREATE_SUBFORM_FIELD_WARNING="You need to create some standard fields before creating a Subform field"
```

joomdonation committed:
Update administrator/components/com_admin/sql/updates/mysql/4.0.0-2021-04-11.sql

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

joomdonation committed:
Update administrator/language/en-GB/com_fields.ini

Co-authored-by: AndySDH <andybadwool@hotmail.it>

richard67 commented on administrator/components/com_admin/sql/updates/postgresql/4.0.0-2021-04-11.sql:
```diff
@@ -0,0 +1,2 @@
+-- Add locked field to extensions table.
```
```suggestion
```
This seems to be a remainder from copy & paste.

joomdonation committed:
Change default value

joomdonation committed:
Update administrator/components/com_admin/sql/updates/postgresql/4.0.0-2021-04-11.sql

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

Quy commented:
> @Quy Could you help me with this:
> 
> > As mentioned before, I'd put the 'Subform Only' badge on the side of the field name, instead of under it (similar to how it's done for hidden Menu Items)

Use `span` instead of `div` and move the code near the field name.

joomdonation committed:
Move badge next to Field name

joomdonation committed:
Merge branch 'subfield_ux_draft' of https://github.com/joomdonation/joomla-cms into subfield_ux_draft

joomdonation committed:
CS

richard67 commented on administrator/components/com_fields/tmpl/fields/default.php:
```diff
@@ -146,6 +146,11 @@
 													<?php echo Text::sprintf('JGLOBAL_LIST_NAME_NOTE', $this->escape($item->name), $this->escape($item->note)); ?>
 												<?php endif; ?>
 											</span>
+											<?php if ($item->only_use_in_subform) : ?>
+                                                <div class="small badge bg-secondary">
+													<?php echo Text::_('COM_FIELDS_FIELD_ONLY_USE_IN_SUBFORM_BADGE'); ?>
+                                                </div>
+											<?php endif; ?>
```
Strange code style ... please use tabs to indent, and use enough of them.

AndySDH commented:
Oh, just noticed: The option "Only Use in Subform" should not display in fields of type 'subfields'. As we can't nest a subform inside a subform.

joomdonation commented:
@AndySDH I noticed that this afternoon. I was thinking about using showon but for some reasons, it was not working when I tried this afternoon. Will try to look at it again later.

joomdonation committed:
Do not show Only Use In Subform for subfields type

joomdonation committed:
Merge branch '4.0-dev' into subfield_ux_draft

joomdonation commented:
@AndySDH I implemented latest change request from you. I also updated testing instructions and mark this PR as ready to review. Please help testing it when you have time

Quy commented on administrator/language/en-GB/com_fields.ini:
```diff
@@ -82,6 +85,8 @@ COM_FIELDS_GROUPS_FILTER_SEARCH_LABEL="Search Field Groups"
 COM_FIELDS_GROUPS_TABLE_CAPTION="Table of Field Groups"
 COM_FIELDS_MUSTCONTAIN_A_TITLE_FIELD="Field must have a title."
 COM_FIELDS_MUSTCONTAIN_A_TITLE_GROUP="Field Group must have a title."
+COM_FIELDS_NO_FIELDS_TO_CREATE_SUBFORM_FIELD_WARNING="You need to create some standard fields before creating a Subform field"
```
```suggestion
COM_FIELDS_NO_FIELDS_TO_CREATE_SUBFORM_FIELD_WARNING="You need to create standard fields before creating a Subform field."
```

Commit:
Update administrator/language/en-GB/com_fields.ini

Quy commented on administrator/language/en-GB/plg_fields_subfields.ini:
```diff
@@ -6,6 +6,7 @@
 PLG_FIELDS_SUBFIELDS="Fields - Subform"
 PLG_FIELDS_SUBFIELDS_LABEL="Subform (%s)"
 PLG_FIELDS_SUBFIELDS_PARAMS_CUSTOMFIELD_LABEL="Field"
+PLG_FIELDS_SUBFIELDS_PARAMS_MAX_ROWS_LABEL="Maximum Number Rows"
```
```suggestion
PLG_FIELDS_SUBFIELDS_PARAMS_MAX_ROWS_LABEL="Maximum Rows"
```

Commit:
Update administrator/language/en-GB/plg_fields_subfields.ini

AndySDH commented:
I have tested this item :white_check_mark: successfully on 56c827efe517262a9f1a0a78585c090a7dfcd6aa<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33096">issues.joomla.org/tracker/joomla-cms/33096</a>.</sub>

Quy commented:
Installed `Blog Sample Data`:

> PHP Notice:  Undefined index: only_use_in_subform in \administrator\components\com_fields\src\Model\FieldModel.php on line 171

Quy commented:
For Subform type, hide `Only Use In Subform`?

joomdonation commented:
@Quy 

> PHP Notice: Undefined index: only_use_in_subform in \administrator\components\com_fields\src\Model\FieldModel.php on line 171

When do you see that error? When you edit the field or save it? 

> For Subform type, hide Only Use In Subform?

Yes. It is working like that. I think you see this issue because you applied the PR https://github.com/joomla/joomla-cms/pull/33200 . The code in current PR checks for Subfields, why the applied PR changed it to Subform. I will wait for that PR merged, then make the change and we can test it again. Thanks for reviewing and testing.

Quy commented:
In PHP error log. The demo installation stops at step 1.

I installed the download package without applying #33200.

joomdonation commented:
> In PHP error log. The sample data installation stops at step 1.
> 
> I installed the download package without applying #33200.

Ah, that explained why I could not install Blog Sample Data. I only think about field created using UI (which only_use_in_subform is always available in $data array). I can make a small modification to solve this issue.

joomdonation committed:
Prevent warning

joomdonation committed:
Merge branch '4.0-dev' into subfield_ux_draft

joomdonation committed:
Change language items cause by merge conflict

alikon commented on administrator/components/com_admin/sql/updates/mysql/4.0.0-2019-04-15.sql:
```diff
@@ -1,2 +1,2 @@
 INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`, `client_id`, `enabled`, `access`, `protected`, `manifest_cache`, `params`, `custom_data`, `checked_out`, `checked_out_time`, `ordering`, `state`) VALUES
-(0, 'plg_fields_subfields', 'plugin', 'subfields', 'fields', 0, 1, 1, 0, '', '', '', 0, '0000-00-00 00:00:00', 0, 0);
+(0, 'plg_fields_subform', 'plugin', 'subform', 'fields', 0, 1, 1, 0, '', '', '', 0, '0000-00-00 00:00:00', 0, 0);
```
shouldn't this  '0000-00-00 00:00:00' be NULL ?

richard67 replied:
@alikon No, that file runs before the null date conversion files run, we had that discussion in past. that file is from 2019-04-15.

joomdonation committed:
Field type is now subform

Quy commented:
Edit a subform field. Applied #33262.

![33096](https://user-images.githubusercontent.com/368084/115896104-ed70f600-a40f-11eb-8659-85f226e3923d.jpg)

kiki-G commented:
Getting errors testing 339096
I am getting this error The file marked for modification does not exist: administrator/components/com_fields/src/Field/SubfieldsField.php<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33096">issues.joomla.org/tracker/joomla-cms/33096</a>.</sub>

kiki-G commented:
Getting errors testing 33096
I am getting this error The file marked for modification does not exist: administrator/components/com_fields/src/Field/SubfieldsField.php<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33096">issues.joomla.org/tracker/joomla-cms/33096</a>.</sub>

joomdonation commented:
@kiki-G If it is possible, please setup a fresh installation (generated by this PR) and test it

You can download the full package for this PR here https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/33096/downloads/42296/Joomla_4.0.0-beta8-dev+pr.33096-Development-Full_Package.zip 

Thanks

TLWebdesign commented:
I have tested this item :white_check_mark: successfully on a0dbc206c666d96abf3bd351d8b235f69cc5fd83<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33096">issues.joomla.org/tracker/joomla-cms/33096</a>.</sub>

joomdonation committed:
Fix Only Use In Subform on field editing

joomdonation committed:
Better code

joomdonation commented:
@Quy My last commit should fix your last issue (with field editing). Please help running a new test/

joomdonation committed:
CS

richard67 commented on administrator/components/com_fields/src/Model/FieldModel.php:
```diff
@@ -559,6 +571,24 @@ public function getForm($data = array(), $loadData = true)
+				// Field type could not be changed, so remove showon attribute to avoid js errors
+				$form->setFieldAttribute('only_use_in_subform', 'showon', '');
+			}
+
```
```suggestion
```
Fix PHPCS.

richard67 committed:
PHP CS

Quy commented on administrator/components/com_fields/src/Model/FieldModel.php:
```diff
@@ -559,6 +571,22 @@ public function getForm($data = array(), $loadData = true)
+				// Field type could not be changed, so remove showon attribute to avoid js errors
+				$form->setFieldAttribute('only_use_in_subform', 'showon', '');
+			}
+
```
```suggestion
		}
```

Commit:
Add closing }

Quy commented on administrator/components/com_fields/src/Model/FieldModel.php:
```diff
@@ -559,6 +571,22 @@ public function getForm($data = array(), $loadData = true)
+				// Field type could not be changed, so remove showon attribute to avoid js errors
+				$form->setFieldAttribute('only_use_in_subform', 'showon', '');
+			}
+		}
```
```suggestion
		}

```

Commit:
CS

Quy commented on administrator/components/com_fields/src/Model/FieldModel.php:
```diff
@@ -559,6 +571,23 @@ public function getForm($data = array(), $loadData = true)
+
+			if ($fieldType->value == 'subform')
+			{
+				// Only Use In subform should not is available for subform field type, so we remove it
```
```suggestion
				// Only Use In subform should not be available for subform field type, so we remove it
```

Commit:
Fix comment

AndySDH commented on administrator/components/com_fields/src/Model/FieldModel.php:
```diff
@@ -163,8 +163,20 @@ public function save($data)
-		$cats = ArrayHelper::toInteger($cats);
+
+		/**
+		 * If the field is only use in subform, set Category to None automatically so that it will only be displayed
```
```suggestion
		 * If the field is only used in subform, set Category to None automatically so that it will only be displayed
```

AndySDH commented on administrator/components/com_fields/src/Model/FieldModel.php:
```diff
@@ -559,6 +571,23 @@ public function getForm($data = array(), $loadData = true)
 			$form->setFieldAttribute('state', 'filter', 'unset');
 		}
 
+		// In case are editing a field, field type could not be changed, so some extra handle below is needed
```
```suggestion
		// In case we are editing a field, field type cannot be changed, so some extra handling below is needed
```

richard67 committed:
Fix language in comments (1)

Co-authored-by: AndySDH <andybadwool@hotmail.it>

richard67 committed:
Fix language in comments (2)

Co-authored-by: AndySDH <andybadwool@hotmail.it>

RickR2H commented:
I have tested this item :white_check_mark: successfully on a0dbc206c666d96abf3bd351d8b235f69cc5fd83<br /><br />Installed https://ci.joomla.org/artifacts/joomla/joomla-cms/4.0-dev/33096/downloads/42296/Joomla_4.0.0-beta8-dev+pr.33096-Development-Full_Package.zip and everything works!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33096">issues.joomla.org/tracker/joomla-cms/33096</a>.</sub>

Quy commented:
I have tested this item :white_check_mark: successfully on a0dbc206c666d96abf3bd351d8b235f69cc5fd83<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33096">issues.joomla.org/tracker/joomla-cms/33096</a>.</sub>

Quy commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/33096">issues.joomla.org/tracker/joomla-cms/33096</a>.</sub>

AndySDH commented:
@joomdonation One thing I figured out is that we missed one case.
The Users Component also has custom fields, however, they have no "Categories".

So if you create a field in the Users Component, and you mark it as "Only Use In Subform", nothing is done with that setting, and it still remains in the form when you go to edit user profile (either from backend or frontend).

So to fix this, in this scenario, the field should be removed from the user form. @joomdonation let me know

AndySDH commented:
@joomdonation can you please look into my latest comment here?

AndySDH commented:
@joomdonation pinging you again

joomdonation commented:
@AndySDH Will look at it later. Busy these days to have all my own extensions compatible with J4. Today is the last day :).

AndySDH commented:
Sure, thanks!

AndySDH commented:
@joomdonation any update on my latest comment from May 16, 2021? This is still an issue. Can you please manage to fix this?
